jlm/llvm/opt/alias-analyses/LazyCycleDetection.hpp

```diff
@@ -0,0 +1,187 @@
+      return std::nullopt;
+
+    NumCycleDetectAttempts_++;
+
+    JLM_ASSERT(DfsStack_.empty());
+    DfsStack_.push(superset);
+
+    // Reset all node states
+    std::fill(NodeStates_.begin(), NodeStates_.end(), NodeStateNotVisited);
+
+    while (!DfsStack_.empty())
+    {
+      auto node = DfsStack_.top();
+      if (NodeStates_[node] == NodeStateNotVisited)
+      {
+        NodeStates_[node] = NodeStateVisited;
+        // Make sure all successors get visited
+        for (auto successor : GetSuccessors_(node))
+        {
+          auto successorRoot = Set_.GetUnificationRoot(successor);
+
+          // Cycle found! Do not add the subset to the dfs stack
+          if (successorRoot == subset)
+            continue;
+
+          if (NodeStates_[successorRoot] != NodeStateNotVisited)
+            continue;
+
+          DfsStack_.push(successorRoot);
+        }
+      }
+      else if (NodeStates_[node] == NodeStateVisited)
+      {
+        DfsStack_.pop();
+        NodeStates_[node] = NodeStateLeft;
+
+        // Check if any successors are unified with the subset. If so, join it!
+        for (auto successor : GetSuccessors_(node))
+        {
+          auto successorRoot = Set_.GetUnificationRoot(successor);
+          if (successorRoot == subset)
+          {
+            subset = UnifyPointerObjects_(node, subset);
+            NumCycleUnifications_++;
+            break;
+          }
+        }
+      }
+      else
+      {
+        // The node has already been visited for a second time
+        DfsStack_.pop();
+      }
+    }
+
+    JLM_ASSERT(Set_.IsUnificationRoot(subset));
+    superset = Set_.GetUnificationRoot(superset);
+    if (subset == superset)
+    {
+      NumCyclesDetected_++;
+      return subset;
+    }
+    return std::nullopt;
+  }
+
+  /**
+   * @return the number of DFSs performed to look for cycles
+   */
+  [[nodiscard]] size_t
+  NumCycleDetectionAttempts()
+  {
+    return NumCycleDetectAttempts_;
+  }
+
+  /**
+   * @return the number of cycles detected by Lazy cycle detection
+   */
+  [[nodiscard]] size_t
+  NumCyclesDetected()
+  {
+    return NumCyclesDetected_;
+  }
+
+  /**
+   * @return the number of unifications made while eliminating found cycles
+   */
+  [[nodiscard]] size_t
+  NumCycleUnifications()
+  {
+    return NumCycleUnifications_;
+  }
+
+private:
+  PointerObjectSet & Set_;
+  const GetSuccessorsFunctor & GetSuccessors_;
+  const UnifyPointerObjectsFunctor & UnifyPointerObjects_;
+
+  // A set of all checked simple edges first -> second, to avoid checking again
+  util::HashSet<std::pair<PointerObjectIndex, PointerObjectIndex>> CheckedEdges_;
+
+  // The dfs stack, which may contain the same node multiple times
+  std::stack<PointerObjectIndex> DfsStack_;
+  // Possible states of nodes during the DFS
+  static constexpr uint8_t NodeStateNotVisited = 0;
+  static constexpr uint8_t NodeStateVisited = 1;
+  static constexpr uint8_t NodeStateLeft = 2;
```

Comment:
What does the state "NodeStateLeft" mean? Left as in opposite of right, or as in remaining?